File: apps/mo2-android/android/feature/tasks/src/test/kotlin/jp/developershub/dub/mo2/feature/tasks/TaskRepositoryTest.kt
// TaskRepository unit tests (§7) — optimistic status change: commit on 200,
// rollback + refetch on 409, rollback on other error, and the not-in-cache guard.
// Mirrors task-repository.test.ts.
package jp.developershub.dub.mo2.feature.tasks

import jp.developershub.dub.mo2.core.model.TaskStatus
import jp.developershub.dub.mo2.core.network.AppError
import jp.developershub.dub.mo2.core.network.AppErrorException
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class TaskRepositoryTest {

    @Test
    fun `commits authoritative server state on success`() = runTest {
        val fake = FakeMobileBffClient(
            onUpdate = { id, status, _ -> task(id, status, version = 2) },
        )
        val repo = TaskRepository(fake)
        repo.seed(listOf(summary("t1", TaskStatus.TODO)))

        val result = repo.updateStatus("t1", TaskStatus.DONE, baseVersion = 1)

        assertTrue(result is TaskUpdateResult.Ok)
        assertEquals(TaskStatus.DONE, repo.tasks.value.single().status)
        assertEquals(1, fake.updateCalls)
    }

    @Test
    fun `rolls back and refetches on 409 conflict`() = runTest {
        val fake = FakeMobileBffClient(
            onUpdate = { _, _, _ -> throw AppErrorException(AppError.Conflict(serverVersion = 7)) },
            onGetTask = { id -> task(id, TaskStatus.IN_PROGRESS, version = 7) },
        )
        val repo = TaskRepository(fake)
        repo.seed(listOf(summary("t1", TaskStatus.TODO)))

        val result = repo.updateStatus("t1", TaskStatus.DONE, baseVersion = 1)

        assertTrue(result is TaskUpdateResult.Conflict)
        assertEquals(7, (result as TaskUpdateResult.Conflict).serverVersion)
        // rolled back off the optimistic DONE, then refetched to server truth
        assertEquals(TaskStatus.IN_PROGRESS, repo.tasks.value.single().status)
        assertEquals(1, fake.getTaskCalls)
    }

    @Test
    fun `rolls back to previous on non-conflict error`() = runTest {
        val fake = FakeMobileBffClient(
            onUpdate = { _, _, _ -> throw AppErrorException(AppError.Network(retryable = true)) },
        )
        val repo = TaskRepository(fake)
        repo.seed(listOf(summary("t1", TaskStatus.TODO)))

        val result = repo.updateStatus("t1", TaskStatus.DONE, baseVersion = 1)

        assertTrue(result is TaskUpdateResult.Failure)
        assertEquals(TaskStatus.TODO, repo.tasks.value.single().status)
        assertEquals(0, fake.getTaskCalls)
    }

    @Test
    fun `keeps rolled-back value when conflict refetch also fails`() = runTest {
        val fake = FakeMobileBffClient(
            onUpdate = { _, _, _ -> throw AppErrorException(AppError.Conflict(serverVersion = null)) },
            onGetTask = { throw AppErrorException(AppError.Network(retryable = true)) },
        )
        val repo = TaskRepository(fake)
        repo.seed(listOf(summary("t1", TaskStatus.TODO)))

        val result = repo.updateStatus("t1", TaskStatus.DONE, baseVersion = 1)

        assertTrue(result is TaskUpdateResult.Conflict)
        assertEquals(TaskStatus.TODO, repo.tasks.value.single().status)
    }

    @Test
    fun `returns failure when task is not in cache`() = runTest {
        val repo = TaskRepository(FakeMobileBffClient())
        val result = repo.updateStatus("ghost", TaskStatus.DONE, baseVersion = 1)
        assertTrue(result is TaskUpdateResult.Failure)
        assertTrue((result as TaskUpdateResult.Failure).error is AppError.Server)
    }

    @Test
    fun `refreshMyTasks replaces cache from server`() = runTest {
        val fake = FakeMobileBffClient(myTasks = listOf(summary("a", TaskStatus.TODO), summary("b", TaskStatus.DONE)))
        val repo = TaskRepository(fake)
        repo.refreshMyTasks()
        assertEquals(listOf("a", "b"), repo.tasks.value.map { it.id })
    }
}

File: apps/mo2-android/test/task-repository.test.ts
import { describe, it, expect, vi } from "vitest";
import { MobileBffClient } from "../src/bff-client";
import { TaskRepository } from "../src/task-repository";
import { InMemorySessionStore } from "../src/session-store";
import { makeMockServer, errorBody } from "./helpers";

function setup(server = makeMockServer()) {
  const store = new InMemorySessionStore();
  store.setSession("tok_1", "r_1");
  const client = new MobileBffClient({ fetchFn: server.fetch, store, onLogout: vi.fn() });
  const repo = new TaskRepository(client);
  repo.seed([{ id: "tsk_1", title: "T1", status: "todo", assigneeId: "usr_me" }]);
  return { repo, server };
}

function fullTask(over: Record<string, unknown> = {}) {
  return {
    id: "tsk_1",
    eventId: "evt_1",
    title: "T1",
    description: null,
    status: "in_progress",
    priority: "medium",
    assigneeId: "usr_me",
    dueAt: null,
    origin: "internal",
    archivedAt: null,
    version: 2,
    createdAt: "2026-08-09T00:00:00Z",
    updatedAt: "2026-08-09T00:00:00Z",
    ...over,
  };
}

describe("TaskRepository — optimistic status change (§6, theme3 D4)", () => {
  it("success commits the authoritative server task", async () => {
    const server = makeMockServer({ status: 200, body: fullTask({ status: "done", version: 2 }) });
    const { repo } = setup(server);

    const states: string[] = [];
    repo.subscribe((tasks) => states.push(tasks[0]!.status));

    const res = await repo.updateStatus("tsk_1", "done", 1);
    expect(res.ok).toBe(true);
    if (res.ok) expect(res.task.version).toBe(2);
    expect(repo.observe()[0]!.status).toBe("done");
    // optimistic "done" emitted first, then committed "done"
    expect(states[0]).toBe("done");
  });

  it("409 -> rollback optimistic UI then refetch latest server state", async () => {
    const server = makeMockServer(
      { status: 409, body: errorBody("TASK_VERSION_CONFLICT", { serverVersion: 5 }) }, // PATCH
      { status: 200, body: fullTask({ status: "blocked", version: 5 }) }, // getTask refetch
    );
    const { repo } = setup(server);

    const seen: string[] = [];
    repo.subscribe((tasks) => seen.push(tasks[0]!.status));

    const res = await repo.updateStatus("tsk_1", "done", 1);
    expect(res.ok).toBe(false);
    if (!res.ok && res.conflict) expect(res.serverVersion).toBe(5);
    // final state = refetched server truth, not the optimistic "done" nor the stale "todo"
    expect(repo.observe()[0]!.status).toBe("blocked");
    // sequence: optimistic done -> rollback todo -> refetched blocked
    expect(seen).toEqual(["done", "todo", "blocked"]);
  });

  it("non-conflict error rolls back to the previous value", async () => {
    const server = makeMockServer({ status: 500, body: errorBody("INTERNAL") });
    const { repo } = setup(server);
    const res = await repo.updateStatus("tsk_1", "done", 1);
    expect(res.ok).toBe(false);
    if (!res.ok && !res.conflict) expect(res.error.kind).toBe("Server");
    expect(repo.observe()[0]!.status).toBe("todo"); // rolled back
  });

  it("updating an id not in cache returns an error without a network call", async () => {
    const server = makeMockServer();
    const { repo } = setup(server);
    const res = await repo.updateStatus("tsk_missing", "done", 1);
    expect(res.ok).toBe(false);
  });
});
